third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,595 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Prerequisites: none.
+//!
+//! Simplifies the control flow graph in the following ways:
+//! (Implemented by `ControlFlowGraphSimplifier`)
+//! 1. Eliminates branch or jump to jump:
+//!     Any empty block
+//!
+//!     L1: goto L2 where L1 != L2
+//!
+//!     is removed, and any block jumping to L1 will jump to L2 directly
+//!
+//!     goto L1
+//!     => (the => here means the instruction above will be replaced by the instruction below)
+//!     goto L2
+//!
+//!     if ... goto L1 else ...
+//!     =>
+//!     if ... goto L2 else ...
+//!
+//!     if ... goto ... else L1
+//!     =>
+//!     if ... goto ... else L2
+//!
+//! 2. Removes edges where the source has only one successor and the target has only one predecessor:
+//!     If there is block
+//!     BB1: xx
+//!          goto L2
+//!     and block
+//!
+//!     BB2: L2
+//!          yy
+//!
+//!     which has only one predecessor BB1.
+//!     Then we move BB2 to the end of BB1 so that BB1 becomes
+//!
+//!     BB1: xx
+//!          yy
+//!
+//!     and BB2 is removed.
+//!
+//! 3. Removes all jumps to the next instruction:
+//!     Whenever we have
+//!
+//!     goto L;
+//!     L;
+//!
+//!     replace it by simply
+//!
+//!     L;
+//!
+//! 4. Replaces branch to same target by jump:
+//!     Whenever we have
+//!
+//!     if ... goto L else goto L
```

Comment:
Wondering whether we can optimize out the if condition after doing this by existing optimizations?

Reply:
I can see one such case: If `if $t goto L else goto L` is transformed into `goto L`, and this is the sole use of `$t`. Then we should be able to optimize `$t` out.